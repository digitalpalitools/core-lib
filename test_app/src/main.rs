use rusqlite::{Connection, Row, NO_PARAMS};

fn get_row_cells(row: &Row) -> Vec<String> {
    let cells: Vec<String> = row
        .column_names()
        .iter()
        .map(|&cn| {
            let cell: String = match row.get(cn) {
                Ok(c) => c,
                Err(e) => e.to_string(),
            };
            cell
        })
        .collect();

    cells
}

fn exec_sql_core(sql: &str) -> rusqlite::Result<Vec<Vec<Vec<String>>>, rusqlite::Error> {
    let conn = Connection::open("./inflections.db")?;
    let mut result: Vec<Vec<Vec<String>>> = Vec::new();
    for s in sql.split(';').filter(|s| !s.trim().is_empty()) {
        let mut stmt = conn.prepare(&s)?;
        let mut rows = stmt.query(NO_PARAMS)?;

        let mut table: Vec<Vec<String>> = Vec::new();
        while let Some(row) = rows.next()? {
            table.push(get_row_cells(row));
        }
        result.push(table)
    }

    Ok(result)
}

fn exec_sql(sql: &str) -> Result<String, String> {
    let table = exec_sql_core(&sql).map_err(|x| x.to_string())?;
    serde_json::to_string(&table).map_err(|x| x.to_string())
}

fn main() -> Result<(), String> {
    println!("{:?}", pls_core::alphabet::PALI_ALPHABET_ROMAN);
    let x = pls_core::alphabet::PaliAlphabet::AA;
    println!("ā > bh? {:#?}", x > pls_core::alphabet::PaliAlphabet::BH);

    let html = pls_core::inflections::generate_inflection_table(
        "vyābādhetvā",
        |s| Ok(s.to_string()),
        exec_sql,
    )?;
    println!("{:#?}", html);

    std::fs::write("d:/delme/inflections.txt", &html).unwrap();

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    // TODO: Need to be abstracted, granularized and moved to pls_core.
    fn basic_inflection_test() {
        let html = pls_core::inflections::generate_inflection_table(
            "ābādheti",
            |s| Ok(s.to_string()),
            exec_sql,
        );

        let approved_html = include_str!("test_data/basic_inflection_test.approved.txt");

        assert_eq!(html.unwrap(), approved_html);
    }
}
