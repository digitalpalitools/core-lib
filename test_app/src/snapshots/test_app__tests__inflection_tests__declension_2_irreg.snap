---
source: test_app/src/main.rs
expression: html

---
<div class="pls-inflection-root">
  <header class="pls-inflection-header"><summary class="pls-inflection-summary">kamma 1 &ndash; "kamma nt" (like )</summary></header>
<table class="pls-inflection-table pls-inflection-type-conjugation">
  <thead>
    <tr>
      <td>kamma_nt</td>
      <td colspan="2">masculine</td>
      <td colspan="2">feminine</td>
      <td colspan="2">neuter</td>
    </tr>
    <tr>
      <td>case</td>
      <td>sg</td>
      <td>pl</td>
      <td>sg</td>
      <td>pl</td>
      <td>sg</td>
      <td>pl</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <p>nom</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kamma</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammaṃ</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammāni</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammā</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>acc</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kamma</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammaṃ</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammāni</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kamme</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>instr</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammena</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammunā</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammanā</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammehi</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>dat</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammassa</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammuno</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammānaṃ</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>abl</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammasmā</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammamhā</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammunā</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammehi</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>gen</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammassa</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammuno</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammānaṃ</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>loc</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kamme</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammasmiṃ</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammamhi</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammani</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammesu</span></div>
        </td>
    </tr>
    <tr>
      <td>
        <p>voc</p>
      </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kamma</span></div>
        </td>
      <td>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammāni</span></div>
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kammā</span></div>
        </td>
    </tr>
    <tr>
      <td>in comps</td>
      <td colspan="6">
        <div class="pls-inflection-inflected-word"><span class="pls-inflection-inflected-word-suffix">kamma</span></div>
        </td>
    </tr>
  </tbody>
</table>

  <footer class="pls-inflection-footer"><a class="pls-inflection-feedback-link" target="_blank" href="https://docs.google.com/forms/d/e/1FAIpQLSdqnYM0_5VeWzkFBPzyxaLqUfKWgNjI8STCpdrx4vX3hetyxw/viewform"><strong>spot a mistake? something missing? fix it here!</strong></a></footer>
</div>

